### PR TITLE
feat(config): new default value for `mempool.max_txs_bytes` parameter

### LINKED ISSUE #2756: mempool/config: reduce the default `max_txs_bytes` value
The current default value is 1GB (1024**3).

Given that the default block size (a consensus parameter, not a config one) was reduced to 4MB, we should adapt the default maximum size in bytes for the mempool accordingly. 

For instance, with maximum-size blocks, the mempool would be drained after 256 blocks by default. A factor of around 10 would be advisable.

_Originally posted by @sergio-mena in https://github.com/cometbft/cometbft/pull/2352#discussion_r1556509571_

### PR DESCRIPTION
**Pull request by cason**
Closes #2756.

The proposed value is 64 MiB = 16 x default `block.max_bytes` (consensus parameter).

---

#### PR checklist

- [ ] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: docs/references/config/config.toml.md [hunk 1]
```diff
@@ -1256,32 +1256,41 @@ If the mempool is full, incoming transactions are dropped.
 
 The value `0` is undefined.
 
-### mempool.max_txs_bytes
-The maximum size of all transactions accepted in the mempool.
+### mempool.max_tx_bytes
```
Comment (cason) [inline docs/references/config/config.toml.md:1259]: Moved up.

File: .changelog/unreleased/features/2756-mempool-max-txs-bytes.md
Comment (cason) [inline .changelog/unreleased/features/2756-mempool-max-txs-bytes.md:1]: Question: we point to the issue or the PR number? In the file name and in its content?

Comment (ancazamfir) [reply .changelog/unreleased/features/2756-mempool-max-txs-bytes.md:1]: imo to the issue

**Comment by ancazamfir** [comment]
> The proposed value is 64 MiB = 16 x default block.max_bytes (consensus parameter).

Looks good.

File: config/config.go [hunk 2]
```diff
@@ -917,9 +917,9 @@ func DefaultMempoolConfig() *MempoolConfig {
 		// Each signature verification takes .5ms, Size reduced until we implement
 		// ABCI Recheck
 		Size:        5000,
-		MaxTxsBytes: 1024 * 1024 * 1024, // 1GB
+		MaxTxBytes:  1024 * 1024,          // 1MB
+		MaxTxsBytes: 16 * 4 * 1024 * 1024, // 64MB
```
Comment (ancazamfir) [inline config/config.go:921]: Curious why not? Trying to show `16 * max_block_size *...`?
```suggestion
		MaxTxsBytes: 64 * 1024 * 1024, // 64MB
```

**Review by ancazamfir — APPROVED** [review]

**Review by melekes — COMMENTED** [review]
`docs/explanation/core/configuration.md` needs to be updated as well

**Comment by hvanz** [comment]
I'd say that 64 Gib is a low number but this is just a guess. If it happens that a few big txs are submitted at the same time, that would take up a big chunk of the mempool capacity and leave a reduced space for the regular-sized txs. Cosmwasm smart contracts could take a couple of Mbs (see https://github.com/cometbft/cometbft/pull/1518#issuecomment-1793240623).

File: config/config.go (hunk 2 above)
Comment (cason) [reply config/config.go:921]: Yes.

**Comment by cason** [comment]
> Cosmwasm smart contracts could take a couple of Mbs

In this case, they have to change the default `max_tx_bytes` (1MB) as well...

Notice that this is not a consensus parameter, is something pretty easy to tune. The real point here is whether it makes sense to allow 1GB transactions by default with block size of 4MB (default again). It would take 256 rounds to drain all transactions from the mempool.

**Comment by cason** [comment]
> `docs/explanation/core/configuration.md` needs to be updated as well

Good point, I will fix it.

But should we maintain this document (as well)? Another source to manually update each time we change a config param...

**Comment by melekes** [comment]
> > `docs/explanation/core/configuration.md` needs to be updated as well
> 
> Good point, I will fix it.
> 
> But should we maintain this document (as well)? Another source to manually update each time we change a config param...

I would probably remove it, yeah. Now that we have `config.toml.md`

**Review by melekes — APPROVED** [review]

File: docs/explanation/core/configuration.md [hunk 3]
```diff
@@ -27,7 +27,7 @@ like the file below, however, double check by inspecting the
 
 # The version of the CometBFT binary that created or
 # last modified the config file. Do not modify this.
-version = "0.39.0"
+version = "1.0.0-dev"
```
Comment (cason) [inline docs/explanation/core/configuration.md:30]:
```suggestion
version = "1.0.0"
```

File: config/config.go (hunk 2 above)
Comment (andynog) [reply config/config.go:921]: Agree, ideally if there’s a correlation with these parameters, e.g. tx size, # txs per block, block size, we should look into initializing some values (e.g. MaxTxsBytes = RoundsToDrain * BlockSize * MaxTxBytes) based on a calculation of things that would constrain it

File: docs/explanation/core/configuration.md (hunk 3 above)
Comment (sergio-mena) [reply docs/explanation/core/configuration.md:30]: We can't set it to exactly "1.0.0" on `main`, as we risk to confuse it with the first final release of `v1.x`

**Review by sergio-mena — APPROVED** [review]
🙏

File: config/config.go (hunk 2 above)
Comment (cason) [reply config/config.go]: Simplified the math, added rationale as a comment

**Comment by cason** [comment]
> I would probably remove it, yeah. Now that we have `config.toml.md`

#2769

File: config/config.go (hunk 2 above)
Comment (andynog) [reply config/config.go]: Thanks !

File: docs/references/config/config.toml.md [hunk 4]
```diff
@@ -1256,32 +1256,41 @@ If the mempool is full, incoming transactions are dropped.
...
-max_tx_bytes = 1048576
+max_txs_bytes = 67108864
 ```
 
 | Value type          | integer |
 |:--------------------|:--------|
 | **Possible values** | &gt;= 0 |
 
-This is the maximum size of a transaction allowed to be accepted into the mempool.
+This is the raw, total transaction size. For example, given 1MB
```
Comment (andynog) [inline docs/references/config/config.toml.md:1283]:
```suggestion
This is the raw, total size in bytes of all transactions in the mempool. For example, given 1MB
```
`total transaction size` might need some better clarification ?

File: docs/references/config/config.toml.md [hunk 5]
```diff
@@ -1256,32 +1256,41 @@ If the mempool is full, incoming transactions are dropped.
...
+transactions and a 5MB maximum mempool byte size, the mempool will
+only accept five transactions.
+
+The maximum mempool byte size should be a factor of the network's maximum block size
+(which is a [consensus parameter](https://docs.cometbft.com/v1.0/spec/abci/abci++_app_requirements#blockparamsmaxbytes)).
+The rationale is to consider how many blocks have to be produced in order to
+drain all transactions stored in a full mempool.
```
Comment (andynog) [inline docs/references/config/config.toml.md:1290]: Should we add a blurb of text saying what happens if the mempool is "full" ? We added that to the max tx size, but probably would be important to also highlight the fact that if the mempool is considered full txs won't be accepted ?

**Review by andynog — APPROVED** [review]
lgtm, added a few suggestions.

**Pull request merged**